### PR TITLE
grayscale filter fix

### LINKED ISSUE #590: Error when creating grayscale pdf
### Describe your issue here. 
Error occures "Error on creating application folder".

### Steps to reproduce
* Create PDF
* Select Images
* Create grayscale PDF.
* Enter Name and Press OK

### Expected behaviour
* PDF should be created correctly
* Images shouldn't become grayscale when `Cancel` is pressed.

### Actual behaviour
* Error occures "Error on creating application folder".
* Shows images count as it is but when we click on any enhancement option Snackbar shows "No images Selected".
* Create PDF remains Enabled.

### PR DESCRIPTION
**Pull request by Ni3verma**
# Description
grayscale image is getting created, no error if we press cancel and other enhancement options are also working fine

Fixes #590 

## Type of change
Just put an x in the [] which are valid.
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.
- [x] `./gradlew assembleDebug assembleRelease`
- [x] `./gradlew checkstyle`

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

**Comment by Ni3verma** [comment]
I am not getting any error. Please tell me steps to reproduce the error

**Comment by vyankatesh24** [comment]
> I am not getting any error. Please tell me steps to reproduce the error

Same as in issue #590

**Comment by Ni3verma** [comment]
No error on my device. Are you running it on emulator? Try again after reinstalling the app.

**Comment by vyankatesh24** [comment]
> No error on my device. Are you running it on emulator? Try again after reinstalling the app.

I have already tried reinstalling.

**Comment by Ni3verma** [comment]
I am not getting any error.
@Swati4star @sidhuparas please check if you are also getting the error

**Comment by vyankatesh24** [comment]
> I am not getting any error.
> @Swati4star @sidhuparas please check if you are also getting the error

Try reinstalling the app. The error comes when you do `grayscale PDF` for the first time. Then after it's working already.

Your solution is solving one point in issue
* Images shouldn't become grayscale when `Cancel` is pressed.

**Comment by Ni3verma** [comment]
Still no error on my device. Please ask other mentors to test on their device.

**Comment by Ni3verma** [comment]
I just encountered the error...but it was totally random.
I am not able to reproduce it again.
Only thing I did different was that I selected 4 images instead of 2

**Comment by sidhuparas** [comment]
@Ni3verma Sorry for the delay, will have a look by evening.

**Comment by sidhuparas** [comment]
@Ni3verma Same. Not working for the first time but then working fine onwards. This is getting repeated on every app restart.

**Comment by vyankatesh24** [comment]
> I just encountered the error...but it was totally random.
> I am not able to reproduce it again.
> Only thing I did different was that I selected 4 images instead of 2

Are you able to find a solution?

**Comment by Ni3verma** [comment]
@vyankatesh24 
trying to find the problem.

**Comment by Ni3verma** [comment]
> @Ni3verma Same. Not working for the first time but then working fine onwards. This is getting repeated on every app restart.

I was launching the app from recent menu, that's why I was not getting this error. thanks for this info :v:

**Comment by Ni3verma** [comment]
@vyankatesh24 @sidhuparas
There is some problem with grayscale transformation of picasso.
See getTarget method in imagesToPdfFragment. It's onBitmapLoaded is not getting called for every image.
I need help here.

**Comment by sidhuparas** [comment]
@Ni3verma I did some logging and got this in `CreatePdf` class's `doInBackground()`: 
`The document has no pages.`
This might help a bit. Also the grayscale photos aren't getting removed from Gallery after operation completion.

**Comment by Ni3verma** [comment]
@sidhuparas I know about this log message. It happens because images are not being converted to grayscale.

**Comment by Ni3verma** [comment]
@sidhuparas @vyankatesh24 @Swati4star 
I need help with this issue

**Comment by vyankatesh24** [comment]
> @sidhuparas @vyankatesh24 @Swati4star
> I need help with this issue

Error is with `getTarget()` method. in `saveCurrentImageInGrayscale`

**Comment by Ni3verma** [comment]
> @vyankatesh24 @sidhuparas
> There is some problem with grayscale transformation of picasso.
> See getTarget method in imagesToPdfFragment. It's onBitmapLoaded is not getting called for every image.
> I need help here.

@vyankatesh24 I already know about it. guide me how to solve it

**Comment by vyankatesh24** [comment]
@Ni3verma 
What you can do is add Transformed URI to different ArrayList.
Don't remove URIs from `mImagesUri` in for loop
Outside for loop check, if both are equal in size.
If yes replace `mImagesUri` ArrayList items
If no reiterate transformation process

File: app/src/main/java/swati4star/createpdf/fragment/ImageToPdfFragment.java [hunk 1]
```diff
@@ -460,23 +461,25 @@ public void onItemClick(int position) {
...
                 String fileName = String.format(getString(R.string.filter_file_name),
-                        String.valueOf(System.currentTimeMillis()), "grayscale");
+                        String.valueOf(System.currentTimeMillis()), i + "_grayscale");
                 File outFile = new File(dir, fileName);
                 String imagePath = outFile.getAbsolutePath();
-                picasso.load(new File(mImagesUri.get(countElements)))
+                picasso.load(new File(mImagesUri.get(i)))
                         .transform(transformation)
                         .into(getTarget(imagePath));
-                mImagesUri.remove(countElements);
```
Comment (Ni3verma) [inline app/src/main/java/swati4star/createpdf/fragment/ImageToPdfFragment.java]: @vyankatesh24 I have already tried this

Comment (vyankatesh24) [reply app/src/main/java/swati4star/createpdf/fragment/ImageToPdfFragment.java]: Outside for loop check, if both are equal in size.
If `yes` replace `mImagesUri` ArrayList items
If `no` reiterate transformation process

Comment (vyankatesh24) [reply app/src/main/java/swati4star/createpdf/fragment/ImageToPdfFragment.java]: @Ni3verma Is this solution working?

Comment (Ni3verma) [reply app/src/main/java/swati4star/createpdf/fragment/ImageToPdfFragment.java]: No. It resulted in 1k transformed photos in my storage but size of arraylists never become equal. I had to force close the app

Comment (Swati4star) [reply app/src/main/java/swati4star/createpdf/fragment/ImageToPdfFragment.java]: Do we have to use a new class variable Arraylist to store temp images?

**Comment by Ni3verma** [comment]
Please review it again, now problem should be solved. It is a super hard issue !

File: app/src/main/java/swati4star/createpdf/fragment/ImageToPdfFragment.java [hunk 2]
```diff
@@ -456,31 +462,69 @@ public void onItemClick(int position) {
...
+            e.printStackTrace();
         }
     }
 
+    /**
+     * convert a bitmap to grayscale and return it
+     * @param bmpOriginal original bitmap which is converted to a new
+     *                    grayscale bitmap
+     */
+    private Bitmap toGrayscale(Bitmap bmpOriginal) {
```
Comment (Swati4star) [inline app/src/main/java/swati4star/createpdf/fragment/ImageToPdfFragment.java]: you can add this function to imageutils file

**Comment by vyankatesh24** [comment]
I've tested the changes, the issue is solved.
Well Done @Ni3verma :+1:

File: app/src/main/java/swati4star/createpdf/fragment/ImageToPdfFragment.java (hunk 2 above)
Comment (Ni3verma) [reply app/src/main/java/swati4star/createpdf/fragment/ImageToPdfFragment.java]: review the changes again

**Comment by Swati4star** [comment]
@vyankatesh24 Please merge this

**Pull request merged**